### ISSUE TITLE
[Question]: what antlr directory really do? 

### ISSUE THREAD

**Issue by louishust** [babelfish-for-postgresql/babelfish_extensions#47]
### Contact Details

hdchild@163.com

### What's the question?

I have a question!

I see there is a directory named antlr under babelfishpg_tsql, but babelfishpg_tsql generate gram-backend.y for parser,
so what the antlr really do?

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by korry** [comment]
It's not uncommon for a procedural language to have a parser (src/backend/pl separate from the backend (SQL) parser (src/backend/parser/gram.y).

The antlr directory holds the parser (and lexer) for PL/tsql - this parser (and lexer) isconstructed using ANTLR4.

I think that gram-backend.y is a modified version of src/backend/parser/gram.y - it's used to parse SQL code that may contain syntax extensions from TSQL (for example, SELECT TOP 5).

**Comment by louishust** [comment]
@korry 
Thanks for your reply.
I saw there is an option named `babelfishpg_tsql.use_antlr` to choose the pl/tsql's parser.

```
 879 ▸ if (pltsql_use_antlr)
 880 ▸ {
 881 ▸ ▸ ANTLR_result result = antlr_parser_cpp(proc_source);
 882
 883 ▸ ▸ if (result.success)
 884 ▸ ▸ {
 885 ▸ ▸ ▸ parse_rc = 0;
 886 ▸ ▸ }
 887 ▸ ▸ else
 888 ▸ ▸ {
 889 ▸ ▸ ▸ report_antlr_error(result);
 890 ▸ ▸ ▸ parse_rc = 1; /* invalid input */
 891 ▸ ▸ }
 892 ▸ }
 893 ▸ else
 894 ▸ {
 895 ▸ ▸ parse_rc = pltsql_yyparse();
 896 ▸ }
```
So there are two ways to parse the pltsql? one using antlr and the other use yacc?
and which way will be the primary way in future?

**Comment by gopinathpai** [comment]
@korry - could you please respond to the follow up question?

**Comment by sangilsong** [comment]
Hi @louishust,

The current babelfish architecture has two parsers. The first one is ANTLR parser which is compatible with T-SQL grammar.
This is the frontend parser (but still running postgres backend) and consume all the query string inputed from TDS port.
The role of ANTLR parser is 1) validate T-SQL grammar and then 2) relay the query string into postgres backend parser for normal DML/DDL, or 3) generate a PL/tsql structure which is used to run a T-SQL specific statement like IF, WHILE, ...

This sounds little bit confusing especially that babelfish still relies on PG's backend parser. But, even in a select statement, it takes much efforts to generate execution plan by ANTLR parser directly. so we decided to use PG's backend parser if PG already has similar statement.

But the problem is that, we mainly rely on PG's backend parser for DML/DDL but all the detailed grammar is not same. 
For example, T-SQL has TOP expression but postgres has LIMIT expression. 
To deal with this, we want to add a grammar to accept top-expression in PG, but we want to minimize to touch the PG's backend parser code. gram-backend.y is automatically generated by concatenating PG's gram.y and extended grammar files under babelfishpg_tsql/src/backend_parser. 

Hope that this helps you for better understanding.
And here are the answers for your specific questions.

Q: there are two ways to parse the pltsql? one using antlr and the other use yacc?
A: NO. we need both ANLTR parser and yacc (PG backend parser)

And babelfishpg_tsql.use_antlr will be removed. This was added for historically reason. sorry for confusion of this.
